plutus-core/plutus-core/test/Evaluation/Spec.hs

```diff
@@ -1,11 +1,247 @@
+  _ -> sequenceA $ case meaning of
+    BuiltinMeaning tySch _ _ -> go tySch
+      where
+        go :: forall args res. TypeScheme Term args res -> [Gen (Term, String)]
+        go = \case
+          TypeSchemeResult    -> []
+          TypeSchemeArrow sch -> genArg (typeRep @(Head args)) : go sch
+          TypeSchemeAll _ sch -> go sch
+  where
+    meaning :: BuiltinMeaning Term (CostingPart DefaultUni DefaultFun)
+    meaning = toBuiltinMeaning bn
+
+-- | Generate one argument to a builtin function based on its `TypeRep`.
+genArg :: forall k (a :: k). TypeRep a -> Gen (Term, String)
+genArg tr
+  | Just HRefl <- eqTypeRep tr (typeRep @()) = pure $ mkArg ()
+  | Just HRefl <- eqTypeRep tr (typeRep @Integer) = pure . mkArg =<< genInteger
+  | Just HRefl <- eqTypeRep tr (typeRep @Int) = pure . mkArg =<< genInteger
+  | Just HRefl <- eqTypeRep tr (typeRep @Bool) = pure . mkArg =<< Gen.bool
+  | Just HRefl <- eqTypeRep tr (typeRep @BS.ByteString) = pure . mkArg =<< genByteString
+  | Just HRefl <- eqTypeRep tr (typeRep @Text) = pure . mkArg =<< genText
+  | Just HRefl <- eqTypeRep tr (typeRep @Data) = pure . mkArg =<< genData 5
+  | Just [SomeTypeRep tr1, SomeTypeRep tr2] <- matchTyCon @(,) tr = do
+      (arg1, argStr1) <- genArg tr1
+      (arg2, argStr2) <- genArg tr2
+      case (arg1, arg2) of
+        (Constant _ (Some (ValueOf uni1 val1)), Constant _ (Some (ValueOf uni2 val2))) ->
+          pure
+            ( Constant
+                ()
+                ( someValueOf
+                    (DefaultUniApply (DefaultUniApply DefaultUniProtoPair uni1) uni2)
+                    (val1, val2)
+                ),
+              show (argStr1, argStr2)
+            )
+        _ -> error "genArg: encountered non-Constant term"
+  | Just [SomeTypeRep trElem] <- matchTyCon @[] tr = do
+      (arg, _) <- genArg trElem
+      case arg of
+        Constant _ (Some (ValueOf uniElem (_ :: b))) -> do
+          (args, argStrings) <- unzip <$> (Gen.list (Range.linear 0 10) $ genArg trElem)
+          let valElems :: [b]
+              valElems =
+                args <&> \case
+                  Constant _ (Some (ValueOf _ valElem')) -> unsafeCoerce valElem'
+                  _                                      -> error "genArg: encountered non-Constant term"
+          pure
+            ( Constant () (someValueOf (DefaultUniApply DefaultUniProtoList uniElem) valElems),
+              show argStrings
+            )
+        _ -> error "genArg: encountered non-Constant term"
+  -- Descend upon `Opaque`
+  | Just [_, SomeTypeRep tr'] <- matchTyCon @Opaque tr = genArg tr'
+  -- Descend upon `SomeConstant`
+  | Just [_, SomeTypeRep tr'] <- matchTyCon @SomeConstant tr = genArg tr'
+  -- In the current implementation, all type variables are instantiated
+  -- to `Integer` (TODO: change this).
+  | Just _ <- matchTyCon' "PlutusCore.Builtin.Polymorphism" "TyVarRep" tr =
```

Comment:
Ah I see, this is where it happens!

(Also, why can't you use the `@` version here? Is it because it's not exported?)

Reply:
`TyVarRep` is a type family so we can't write `@TyVarRep` (and it doesn't have `Typeable` anyway).

Reply:
> `TyVarRep` doesn't have Typeable instance due to it being poly-kinded.

I wasn't aware of this restriction, but you can just instantiate the kind, right?

```
>>> :set -XTypeApplications
>>> :set -XKindSignatures
>>> import Type.Reflection
>>> typeRep @(TyVarRep @(*))
TyVarRep *
```

That `TyVarRep` comes from either `Opaque` or `SomeConstant`, both of which restrict the kind of `rep` to `*` (as other kinds don't make sense there).

Reply:
You are right!